app/bot/when.go

```diff
@@ -0,0 +1,134 @@
+package bot
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+)
+
+// When bot is answer on question "when the stream is started".
+type When struct {
+	now     func() time.Time
+	reactOn []string
+}
+
+// NewWhen makes a new When bot.
+func NewWhen(reactOn []string, nowFunc func() time.Time) *When {
+	log.Printf("[INFO] new when bot is started")
+
+	return &When{
+		reactOn: reactOn,
+		now:     nowFunc,
+	}
+}
+
+// Help returns help message
+func (w *When) Help() string {
+	return genHelpMsg(w.ReactOn(), "расписание эфиров Радио-Т")
+}
+
+// OnMessage returns one entry
+func (w *When) OnMessage(msg Message) Response {
+	if !contains(w.ReactOn(), msg.Text) {
+		return Response{}
+	}
+
+	return Response{
+		Text: when(w.now()),
+		Send: true,
+	}
+}
+
+// ReactOn keys
+func (w *When) ReactOn() []string {
+	return w.reactOn
+}
+
+func when(now time.Time) string {
+	const avgDuration = 2 * time.Hour
+	const whenPrefix = "[каждую субботу, 20:00 UTC](https://radio-t.com/online/)"
+	const tooSmallDuration = "пару секунд"
+
+	now = now.UTC()
+	prevStream, nextStream := closestPrevNextWeekdays(now, time.Saturday, 20, 0)
+	diffToPrev := -prevStream.Sub(now)
+	diffToNext := nextStream.Sub(now)
+
+	var whenCountdown string
+	switch {
+	case diffToPrev < avgDuration:
+		whenCountdown = fmt.Sprintf(
+			"\nНачался %s назад. \nСкорее всего еще идет. \nСледующий через %s",
+			humanizeDuration(diffToPrev, tooSmallDuration),
+			humanizeDuration(diffToNext, tooSmallDuration),
+		)
+	case diffToNext >= 0:
+		whenCountdown = fmt.Sprintf(
+			"\nНачнется через %s",
+			humanizeDuration(diffToNext, tooSmallDuration),
+		)
+	}
+
+	return whenPrefix + whenCountdown
+}
+
+// closestPrevNextWeekdays returns closest next `weekday` at `hour`:`minute` after `t`.
+func closestPrevNextWeekdays(t time.Time, weekday time.Weekday, hour, minute int) (time.Time, time.Time) {
```

Comment:
I guess this supposed to be called closestPrevNextWeekend? Or maybe closestPrevNextSat?
Actually, I'm not fully certain what this func supposed to do and how it does it. From the caller side it looks like an attempt to discover times of the prev and next start of the podcast stream, but I don't really understand this implementation. Alternatively I can think on a simple loop adding 6 days to the current day and for back till it gets a pair of Sat

Reply:
This function is pretty easy (but reading this is not so easy). I try to change a function for our concrete purpose with hardcode. I dont like it, but it will be easy to rewrite this function by someone.

I try to explain current behaviour:
- Get an absolute diff in days between current weekday and stream weekday. `daysDiff`
- Add this diff to current timestamp. now.Date() + `daysDiff`. This is date of stream
- Create a new time.Time obj with date of stream, and constant hours of stream
- Also, get the prev (or next) stream date depends on current calculation

Check out [commit](https://github.com/radio-t/super-bot/pull/83/commits/7f273efec9ba2456c2e6b647dddaa9b45de8856d)

Reply:
Unless I missed the point, this is way too much for what it does. Isn't replacing this thing with a trivial "gimme time duration from now till the next show" enough? It will get rid of prev/next thing completely and will give you back directly what you are looking for - duration till the next show.

```
func(w *When) tillNextShow(t time.Time) (time.Duration) {
	next := t.AddDate(0, 0, int(time.Saturday-t.Weekday()))
	next = time.Date(next.Year(), next.Month(), next.Day(), 20, 0, 0, 0, time.UTC)
	return t.Sub(next)
}
```

Reply:
done